### DISCUSSION TITLE
Strange behaviour Calendar field

### DISCUSSION THREAD

**Discussion by marcorensch** [joomengine/Joomla-Component-Builder#672] [Extention Issues] [answered]
Hi at all, i'm currently struggeling while creating a working datetime field for my component.
In fact it "reverses" the given date and saves it in a totaly wrong string order inside the db:
What i select in the calendar:
![Bildschirmfoto 2021-02-16 um 10 14 16](https://user-images.githubusercontent.com/11912135/108042549-3dae8f00-7040-11eb-9c6e-222393d42384.jpg)
What i have after save:
![Bildschirmfoto 2021-02-16 um 10 14 27](https://user-images.githubusercontent.com/11912135/108042579-47d08d80-7040-11eb-8de4-34095652cab5.jpg)
How its saved in the db:
![Bildschirmfoto 2021-02-16 um 10 14 43](https://user-images.githubusercontent.com/11912135/108042606-5159f580-7040-11eb-80eb-4a7f5dc6a3ca.jpg)

The field format should be:  %d.%m.%Y (German DE format)
My field settings are actually:
default: NOW
format: true (should not been used because see below)
filter: STRING
translateformat: true
showtime: false

For me it looks like the field does not adapt the visible format to the SQL format and puts the numbers in as given (for sure that was not the case before) - i have already tried to complete the affected fields and re-created them from scratch. I've tried to disable the translateformat and used hardcoded format in field settings but the effect still remains the same...

Am i just to dumb or is there a bug?

**Comment by Llewellynvdm** [comment, marked as answer]
The date field is one of those fields that need extra help at times. JCB does nothing to model this field, it is all Joomla and if the outcome is not what you expected you should use the expert modeling option in the field to model the field yourself.
![image](https://user-images.githubusercontent.com/5607939/108130546-82760c80-70b8-11eb-9b11-cfb1c6fd9684.png)

I have seen other developers face the same issue and they had very hard workarounds. But the expert modeling option gives you the ability to model the field before you save it and again when you retrieve it (if need be) from the database.

So this is what I used at times:
![image](https://user-images.githubusercontent.com/5607939/108130913-16e06f00-70b9-11eb-9d83-0c1124f00066.png)
```php
if (isset([[[field]]]) && ([[[field]]] != '0000-00-00 00:00:00') && ([[[field]]] != ''))
{
	[[[field]]] = JFactory::getDate([[[field]]])->toSql();
}
```

Then my field setup looks like this:
![image](https://user-images.githubusercontent.com/5607939/108130992-3e373c00-70b9-11eb-8f06-bd08280f3983.png)
![image](https://user-images.githubusercontent.com/5607939/108131014-4b542b00-70b9-11eb-9b41-d66e925c7708.png)

**Reply by Llewellynvdm** [reply]
Since the get modeling will want something... this is what I add:
![image](https://user-images.githubusercontent.com/5607939/108131319-c4538280-70b9-11eb-8e1b-644ec70b2493.png)

**Reply by marcorensch** [reply]
i know there was a video about that but i did not find it currently, When i set the on Save model as you described it is also used in the backend export, there where an if statement you have used in one of your videos if its not export. Can you please point me into the right direction?

**Edit** i found the vid: https://www.youtube.com/watch?v=fau5mZ6naLc&list=PLQRGFI8XZ_wtGvPQZWBfDzzlERLQgpMRE&index=28

and i tried this, but its not working as expected:

```
if(!isset($_export) || !$_export){
if (isset([[[field]]]) && ([[[field]]] != '0000-00-00 00:00:00') && ([[[field]]] != ''))
{
	[[[field]]] = JFactory::getDate([[[field]]])->toSql();
}
}else{
[[[field]]] = [[[field]]];
}
```

**Reply by marcorensch** [reply]
Solved now:
I had in the Admin View Field Relations setted a rule to overwrite the date display, so i've changed it there:
```
// [1031]=> Active from
if(!isset($_export) || !$_export){
	$item->{1031} = Joomla\CMS\HTML\HTMLHelper::date(
		$item->{1031}, Joomla\CMS\Language\Text::_('DATE_FORMAT_FILTER_DATE')
	);
}else{
	// nothing to change
}
```

Admin Views --> My view --> Fields --> Field relations --> **Model (before modelling)** rule
